ls1intum/Artemis
# `General`: Add categories filter to average exercise scores chart

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Currently, we provide an exercise type filter for the corresponding chart in the course statistics. This PR enhances the filter options by providing filter settings based on categories.
### Description

Currently, we provide both exercise type filters as well as category filters for Charts in Artemis. I changed the implementation and moved the filter logic to two new classes. The components now use those classes in order to realize the filtering of the chart content. 

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with multiple exercises and various categories.
- **Important**: Please create different combinations of categories and exercise types. Also, for some exercises, assign more than one category

**Alternative: Use [this test course](https://artemis-test1.artemis.in.tum.de/course-management/40) on TS1 with test user 10-15 as instructor.** There is everything set up already
1. Log in to Artemis as Instructor
2. Navigate to the course statistics.
3. The filter dropdown next to the average scores chart should now also provide filter options for categories
4. Play around with the filter. Make sure that the chart content changes appropriately.
5. Log out and log in as student
6. Navigate to the student statistics of the course
7. Make sure that the general category filter, as well as the type filter for the line chart, still works as expected.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| CourseStatisticsAverageScore.java | 100% | 100% |
| StatisticsService.java | 60% | 91% |
| course-statistics.component.ts | 84.68% | 95.12% |
| exercise-scores-chart.component.ts | 78.94% | 95.19% |
| chart-category-filter.ts | 100% | 100% |
| chart-exercise-type-filter.ts | 100% | 100% |
| chart-filter.ts | 100% | 100% |
| statistics-average-score-graph.component.ts | 74.07% | 94.33% |


### Screenshots


Dark mode:
![image](https://user-images.githubusercontent.com/80622272/175564054-11f4ba00-5a5c-4860-ac61-df69c4da5921.png)

Light mode:
![image](https://user-images.githubusercontent.com/80622272/175564149-aee9cb57-f93b-40b6-a07a-1ef02b0622a0.png)
